repos/system_upgrade/common/actors/reportleftoverpackages/reportleftoverpackages/tests/test_reportleftoverpackages.py

```diff
@@ -0,0 +1,45 @@
+from leapp import reporting
+from leapp.libraries.actor import reportleftoverpackages
+from leapp.libraries.common.testutils import create_report_mocked, CurrentActorMocked, logger_mocked
+from leapp.libraries.stdlib import api
+from leapp.models import LeftoverPackages, RemovedPackages, RPM
+
+
+def test_no_leftover_and_no_removed_packages(monkeypatch):
+    monkeypatch.setattr(api, 'current_actor', CurrentActorMocked(msgs=[]))
+    monkeypatch.setattr(reporting, 'create_report', create_report_mocked())
+    monkeypatch.setattr(api, 'current_logger', logger_mocked())
+    reportleftoverpackages.process()
+
+    assert reporting.create_report.called == 0
+    assert api.current_logger.infomsg == ['No leftover packages, skipping...']
+
+
+def test_no_removed_packages_leftover_present(monkeypatch):
+    leftover_packages = LeftoverPackages(items=[RPM(name='rpm', version='1.0', release='1.el7', epoch='0',
+                                                    packager='foo', arch='noarch', pgpsig='SIG')])
+    monkeypatch.setattr(api, 'current_actor', CurrentActorMocked(msgs=[leftover_packages]))
+    monkeypatch.setattr(reporting, 'create_report', create_report_mocked())
+    monkeypatch.setattr(api, 'current_logger', logger_mocked())
+    reportleftoverpackages.process()
+
+    assert reporting.create_report.called == 1
+    assert 'Some RHEL packages have not been upgraded' in reporting.create_report.report_fields['title']
+    assert 'Following RHEL packages have not been upgraded' in reporting.create_report.report_fields['summary']
+    summary = 'Please remove these packages to keep your system in supported state.'
+    assert summary in reporting.create_report.report_fields['summary']
+    assert 'rpm-1.0-1.el7' in reporting.create_report.report_fields['summary']
+
+
+def test_removed_packages(monkeypatch):
+    removed_packages = RemovedPackages(items=[RPM(name='rpm', version='1.0', release='1.el7', epoch='0',
+                                                  packager='foo', arch='noarch', pgpsig='SIG')])
+
+    monkeypatch.setattr(api, 'current_actor', CurrentActorMocked(msgs=[removed_packages]))
```

Comment:
Looking at the actors producing LeftoverPackages and RemovedPackages, they always produce the message, just with an empty `items` list, if they are not produced, they (or previous actors) probably failed with an error. I would include the empty messages here to better reflect the expected inputs/outputs. Or you could test both cases with a simple parametrize.

Reply:
Added also empty messages.